python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,899 @@
+                                    + cnt
+                                )
+
+        async def publish_messages(self) -> None:
+            assert self.redis_client is not None
+            for (
+                sharded_channel,
+                sharded_messages_to_publish,
+            ) in self.sharded_messages_to_publish.items():
+                for message, cnt in sharded_messages_to_publish.items():
+                    for i in range(cnt):
+                        await self.redis_client.publish(
+                            message=message, channel=sharded_channel, sharded=True
+                        )
+
+            for (
+                unsharded_channel,
+                unsharded_messages_to_publish,
+            ) in self.unsharded_messages_to_publish.items():
+                for message, cnt in unsharded_messages_to_publish.items():
+                    for i in range(cnt):
+                        await self.redis_client.publish(
+                            message=message, channel=unsharded_channel, sharded=False
+                        )
+
+    @staticmethod
+    async def publish_and_assert(all_clients: List["TestPubSub.PubSubClient"]):
+        for client in all_clients:
+            await client.publish_messages()
+
+        # TODO - Wait for how long?
+        await asyncio.sleep(1)
+
+        for client in all_clients:
+            client.calculate_expected_messages(all_clients=all_clients)
+            client.assert_client()
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("sharded_pubsub", [True, False])
+    async def test_pubsub_basic(self, request, cluster_mode, sharded_pubsub):
+        CHANNEL_A_NAME = "test-channel-a"
+
+        callback_a_called = False
+
+        redis_client: (
+            RedisClient | RedisClusterClient | TestPubSub.RedisClientPubSubMock | None
+        ) = None
+        if not request.config.getoption("--mock-pubsub"):
+            redis_client = await create_client(request, cluster_mode)
+        else:
+            redis_client = TestPubSub.RedisClientPubSubMock()
+
+        def callback_a(message: CoreCommands.PubSubMsg) -> None:
+            nonlocal callback_a_called
+            nonlocal redis_client
+
+            callback_a_called = True
+            assert message.context == redis_client
+            assert message.channel == CHANNEL_A_NAME
+            assert not message.pattern
+
+        await redis_client.subscribe(
+            channels_or_patterns={CHANNEL_A_NAME},
+            channel_mode=(
+                CoreCommands.ChannelModes.Sharded
+                if sharded_pubsub
+                else CoreCommands.ChannelModes.Exact
+            ),
+            callback=callback_a,
+            context=redis_client,
+        )
+
+        await redis_client.publish(
+            message="hi", channel=CHANNEL_A_NAME, sharded=sharded_pubsub
+        )
+        await asyncio.sleep(1)
```

Comment:
from past experience: these pubsub tests with await tend to be flakey. Instead, can you wait on a future and pass it as a context?